File: C++/ScrambleString.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
using namespace std;
class Solution {
public:
    bool isScramble(string s1, string s2) {
		// Start typing your C/C++ solution below
		// DO NOT write int main() function
		if(s1 == s2)
			return true;
		int len1 = s1.size();
		int len2 = s2.size();
		if(len1 != len2)
			return false;
		for(int i = 1; i <= len1/2; i++)
		{
			string a1 = s1.substr(0,i);
			string b1 = s1.substr(i);
			string a2 = s2.substr(0,i);
			string b2 = s2.substr(i);
			string c1 = s1.substr(0,len1-i);
			string d1 = s1.substr(len1-i);
			string c2 = s2.substr(0,len1-i);
			string d2 = s2.substr(len1-i);
			if(isScramble(a1, a2) && isScramble(b1, b2))
				return true;
			if(isScramble(a1, d2) && isScramble(b1, c2))
				return true;
			if(isScramble(d1, d2) && isScramble(c1, c2))
				return true;
			if( isScramble(d1, a2) && isScramble(c1, b2))
				return true;
		}
		return false;
	}
};

File: JAVA/ScrambleString.java
class Solution {
    public boolean isScramble(String s1, String s2) {
        // Start typing your Java solution below
        // DO NOT write main() function
    	if(s1.equals(s2))
    		return true;
        int len1 = s1.length();
        int len2 = s2.length();
        if(len1 != len2)
        	return false;
        else if(len1 == 0 || len2 == 0)
        	return false;
        for(int i = 1; i <= len1/2; i++)
        {
        	String a1 = s1.substring(0,i);
        	String b1 = s1.substring(i);
        	String a2 = s2.substring(0,i);
        	String b2 = s2.substring(i);
        	String c1 = s1.substring(0,len1-i);
        	String d1 = s1.substring(len1-i);
        	String c2 = s2.substring(0,len1-i);
        	String d2 = s2.substring(len1-i);
        	if(isScramble(a1, a2) && isScramble(b1, b2))
        		return true;
        	if(isScramble(a1, d2) && isScramble(b1, c2))
        		return true;
        	if(isScramble(d1, a2) && isScramble(c1, b2))
        		return true;
        	if(isScramble(d1, d2) && isScramble(c1, c2))
        		return true;
        }
        return false;
    }
}
public class ScrambleString {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}
